File: src/solana_config/transformation/transaction.rs
use crate::solana_config::proto_codegen::{confirmed_block::CompiledAccount, etl_block::EtlBlock};

#[cfg(feature = "STRING_TIMESTAMP")]
use {
    crate::solana_config::proto_codegen::records_string_timestamp::{
        BalanceChangeRecord, InstructionRecord, TokenBalanceRecord, TokenTransferRecord,
        TransactionAccountRecord, TransactionRecord,
    },
    chrono::{NaiveDateTime, TimeZone, Utc},
};

#[cfg(feature = "INT_TIMESTAMP")]
use crate::solana_config::proto_codegen::records_int_timestamp::{
    BalanceChangeRecord, InstructionRecord, TokenBalanceRecord, TokenTransferRecord,
    TransactionAccountRecord, TransactionRecord,
};

use super::{instruction, token_transfer};

#[derive(Clone)]
pub struct CommonTableFields {
    pub block_slot: Option<i64>,
    pub block_hash: Option<String>,
    #[cfg(feature = "STRING_TIMESTAMP")]
    pub block_timestamp: Option<String>,
    #[cfg(feature = "INT_TIMESTAMP")]
    pub block_timestamp: Option<i64>,
    pub tx_signature: Option<String>,
}

pub fn transform_to_transaction_records(
    etl_block: &EtlBlock,
) -> (
    Vec<TransactionRecord>,
    Vec<InstructionRecord>,
    Vec<TokenTransferRecord>,
) {
    let mut transaction_records = Vec::with_capacity(etl_block.transactions.len());
    let mut instruction_records = Vec::new();
    let mut token_transfer_records = Vec::new();

    let table_context = etl_block.table_context.to_owned().unwrap();
    let block_slot = Some(etl_block.slot as i64);
    let block_hash = Some(table_context.block_hash.to_owned());

    #[cfg(feature = "STRING_TIMESTAMP")]
    let block_timestamp = table_context.block_timestamp.as_ref().map(|bt| {
        Utc.from_utc_datetime(&NaiveDateTime::from_timestamp_opt(bt.timestamp, 0).unwrap())
            .to_rfc3339()
    });
    #[cfg(feature = "INT_TIMESTAMP")]
    let block_timestamp = table_context
        .block_timestamp
        .to_owned()
        .map(|bt| bt.timestamp * 1_000_000);

    for (tx_index, tx_with_meta) in etl_block.transactions.iter().enumerate() {
        let recent_block_hash: Option<String>;
        let tx_lead_signature: Option<String>;
        let transaction_account_records: Vec<TransactionAccountRecord>;
        let tx_accounts: Vec<CompiledAccount>;
        let common_table_fields: CommonTableFields;
        match &tx_with_meta.transaction {
            Some(tx) => {
                tx_lead_signature = tx.signatures.get(0).cloned();
                common_table_fields = CommonTableFields {
                    block_slot,
                    block_hash: block_hash.to_owned(),
                    block_timestamp: block_timestamp.to_owned(),
                    tx_signature: tx_lead_signature.to_owned(),
                };
                match &tx.message {
                    None => {
                        recent_block_hash = None;
                        tx_accounts = Vec::new();
                        transaction_account_records = Vec::new();
                    }
                    Some(tx_message) => {
                        recent_block_hash = Some(tx_message.recent_blockhash.to_owned());
                        tx_accounts = tx_message.account_keys.to_owned();
                        transaction_account_records = tx_accounts
                            .iter()
                            .map(|k| TransactionAccountRecord {
                                pubkey: Some(k.pubkey.to_owned()),
                                signer: Some(k.signer),
                                writable: Some(k.writable),
                            })
                            .collect();

                        // instructions
                        instruction_records.reserve_exact(tx_message.instructions.len());
                        let mut prev_instruction = None;
                        for (instruction_index, instruction) in
                            tx_message.instructions.iter().enumerate()
                        {
                            let instruction_record = instruction::transform_to_instruction_record(
                                common_table_fields.to_owned(),
                                instruction,
                                None,
                                instruction_index,
                            );

                            instruction_records.push(instruction_record);

                            let (instruction_type, token_transfer_instruction) = instruction
                                .parsed_dict
                                .to_owned()
                                .map(|p| (p.r#type, p.token_transfer_instruction))
                                .unwrap_or((None, None));

                            // token transfers
                            if let Some(r#type) = instruction_type.to_owned() {
                                let token_transfer_record =
                                    token_transfer::transform_to_token_transfer_record(
                                        common_table_fields.to_owned(),
                                        &r#type,
                                        instruction.program.to_owned().unwrap_or_default(),
                                        token_transfer_instruction,
                                        prev_instruction,
                                    );

                                if let Some(tt) = token_transfer_record {
                                    token_transfer_records.push(tt);
                                }
                            }
                            prev_instruction = Some(instruction.to_owned());
                        }
                    }
                }
            }
            None => {
                recent_block_hash = None;
                tx_accounts = Vec::new();
                transaction_account_records = Vec::new();
                tx_lead_signature = None;
                common_table_fields = CommonTableFields {
                    block_slot,
                    block_hash: block_hash.to_owned(),
                    block_timestamp: block_timestamp.to_owned(),
                    tx_signature: None,
                };
            }
        }

        match &tx_with_meta.meta {
            Some(meta) => {
                let balance_changes: Vec<BalanceChangeRecord> = tx_accounts
                    .into_iter()
                    .enumerate()
                    .map(|(i, account)| BalanceChangeRecord {
                        account: Some(account.pubkey),
                        before: Some(meta.pre_balances[i]),
                        after: Some(meta.post_balances[i]),
                    })
                    .collect();
                let pre_token_balances = meta
                    .pre_token_balances
                    .iter()
                    .map(|bal| TokenBalanceRecord {
                        account_index: Some(bal.account_index as i64),
                        mint: Some(bal.mint.to_owned()),
                        owner: Some(bal.owner.to_owned()),
                        amount: bal
                            .ui_token_amount
                            .as_ref()
                            .map(|ui_token| ui_token.amount.to_owned()),
                        decimals: bal
                            .ui_token_amount
                            .to_owned()
                            .map(|ui_token| ui_token.decimals as i64),
                    })
                    .collect();
                let post_token_balances = meta
                    .post_token_balances
                    .iter()
                    .map(|bal| TokenBalanceRecord {
                        account_index: Some(bal.account_index as i64),
                        mint: Some(bal.mint.to_owned()),
                        owner: Some(bal.owner.to_owned()),
                        amount: bal
                            .ui_token_amount
                            .as_ref()
                            .map(|ui_token| ui_token.amount.to_owned()),
                        decimals: bal
                            .ui_token_amount
                            .to_owned()
                            .map(|ui_token| ui_token.decimals as i64),
                    })
                    .collect();

                let err = &meta.err;
                let transaction_record = TransactionRecord {
                    block_slot: Some(etl_block.slot as i64),
                    block_hash: Some(table_context.block_hash.to_owned()),
                    block_timestamp: block_timestamp.to_owned(),
                    recent_block_hash,
                    signature: tx_lead_signature.clone(),
                    index: Some(tx_index as i64),
                    fee: Some(meta.fee),
                    status: Some(
                        err.as_ref()
                            .map_or(String::from("Success"), |_| String::from("Failure")),
                    ),
                    err: err.as_ref().map(|e| e.err.to_owned()),
                    compute_units_consumed: meta.compute_units_consumed,
                    accounts: transaction_account_records,
                    log_messages: meta.log_messages.to_owned(),
                    balance_changes,
                    pre_token_balances,
                    post_token_balances,
                };
                transaction_records.push(transaction_record);

                for inner_instructions in &meta.inner_instructions {
                    let parent_index = inner_instructions.index;
                    // this will be updated at the end of each iteration to be the current inner_instruction
                    let mut prev_inner_instruction = None;

                    instruction_records.reserve_exact(inner_instructions.instructions.len());
                    for (inner_index, inner_instruction) in
                        inner_instructions.instructions.iter().enumerate()
                    {
                        let instruction_record = instruction::transform_to_instruction_record(
                            common_table_fields.to_owned(),
                            inner_instruction,
                            Some(parent_index),
                            inner_index,
                        );
                        instruction_records.push(instruction_record);

                        // token transfers
                        let (instruction_type, token_transfer_instruction) = inner_instruction
                            .parsed_dict
                            .to_owned()
                            .map(|p| (p.r#type, p.token_transfer_instruction))
                            .unwrap_or((None, None));
                        if let Some(r#type) = instruction_type.to_owned() {
                            let token_transfer_record =
                                token_transfer::transform_to_token_transfer_record(
                                    common_table_fields.to_owned(),
                                    &r#type,
                                    inner_instruction.program.to_owned().unwrap_or_default(),
                                    token_transfer_instruction,
                                    prev_inner_instruction,
                                );
                            if let Some(tt) = token_transfer_record {
                                token_transfer_records.push(tt);
                            };
                        }
                        prev_inner_instruction = Some(inner_instruction.to_owned());
                    }
                }
            }
            None => {
                let transaction_record = TransactionRecord {
                    block_slot,
                    block_hash: block_hash.clone(),
                    block_timestamp: block_timestamp.to_owned(),
                    recent_block_hash,
                    signature: tx_lead_signature.clone(),
                    index: Some(tx_index as i64),
                    fee: None,
                    status: None,
                    err: None,
                    compute_units_consumed: None,
                    accounts: transaction_account_records,
                    log_messages: Vec::new(),
                    balance_changes: Vec::new(),
                    pre_token_balances: Vec::new(),
                    post_token_balances: Vec::new(),
                };
                transaction_records.push(transaction_record);
            }
        }
    }

    (
        transaction_records,
        instruction_records,
        token_transfer_records,
    )
}

File: storage-write/deprecated/transformation/transaction.go
package transformation

import (
	"encoding/json"
	"fmt"
	"strconv"
	pbcodegen "subscriber_rabbitmq/pbcodegen"
)

func TransformToTransactionRecords(rawBlock *pbcodegen.EtlBlock) ([]*pbcodegen.TransactionRecord, []*pbcodegen.InstructionRecord, []*pbcodegen.TokenTransferRecord) {
	var TransactionRecords []*pbcodegen.TransactionRecord
	var InstructionRecords []*pbcodegen.InstructionRecord
	var TokenTransferRecords []*pbcodegen.TokenTransferRecord

	blockContext := rawBlock.TableContext

	blockSlot := int64(rawBlock.Slot)
	blockHash := blockContext.BlockHash

	var blockTimestamp *int64
	if rawBlock.TableContext.BlockTimestamp != nil {
		_timestamp := rawBlock.TableContext.BlockTimestamp.Timestamp * 1e6
		blockTimestamp = &_timestamp
	}

	for transaction_index, transaction := range rawBlock.Transactions {
		txLeadSignature := ""
		var recentBlockHash *string
		var transaction_accounts []*pbcodegen.CompiledAccount
		var transactionAccountRecords []*pbcodegen.TransactionAccountRecord

		if transaction.Transaction != nil {
			transaction_raw := transaction.Transaction

			if signatures := transaction_raw.GetSignatures(); len(signatures) > 0 {
				txLeadSignature = signatures[0]
			}
			transaction_message := transaction_raw.GetMessage()

			if transaction_message != nil {
				transaction_accounts = transaction_message.GetAccountKeys()
				recentBlockHash = &transaction_message.RecentBlockhash
			} else {
				recentBlockHash = nil
			}

			for _, account := range transaction_accounts {
				transactionAccountRecord := pbcodegen.TransactionAccountRecord{
					Pubkey:   &account.Pubkey,
					Signer:   &account.Signer,
					Writable: &account.Writable,
				}
				transactionAccountRecords = append(transactionAccountRecords, &transactionAccountRecord)
			}
			if len(transactionAccountRecords) == 0 {
				transactionAccountRecord := pbcodegen.TransactionAccountRecord{
					Pubkey:   nil,
					Signer:   nil,
					Writable: nil,
				}
				transactionAccountRecords = append(transactionAccountRecords, &transactionAccountRecord)
			}

			var prev_instruction *pbcodegen.InnerInstruction
			for instruction_index, instruction := range transaction_message.GetInstructions() {
				if instruction == nil {
					continue
				}
				instruction_parsed_dict := instruction.ParsedDict
				var instruction_parsed *string
				var instruction_type *string
				var paramRecords []*pbcodegen.ParamsRecord
				if instruction_parsed_dict == nil {
					instruction_parsed = instruction.ParsedString
				} else {
					instruction_type = instruction_parsed_dict.Type
					instruction_parsed = instruction_parsed_dict.Parsed

					param_keys_values := instruction_parsed_dict.Info
					if param_keys_values != nil {
//fmt.Println("the param_keys_values is", *param_keys_values)		
				var keys_values_map map[string]interface{}
						err := json.Unmarshal([]byte(*param_keys_values), &keys_values_map)
						if err != nil {
							fmt.Println("Error:", err)
							panic("Unexpected instruction params value")
						}
						if len(keys_values_map) == 0 {
							fmt.Println("the keys and values map is empty")
						}
						for k, v := range keys_values_map {
//fmt.Println("the k is:", k)
kCopy := k // Create a copy of k
							jsonBytes, err := json.Marshal(v)
							if err != nil {
								panic(err)
							}
							jsonString := string(jsonBytes)
//fmt.Println("the v string is:", jsonString)
							param_record := pbcodegen.ParamsRecord{
								Key:   &kCopy,
								Value: &jsonString,
							}
//fmt.Println("the param record is:", param_record)					
		paramRecords = append(paramRecords, &param_record)
						}
					}
				}

				accounts := make([]string, len(instruction.Accounts))
				copy(accounts, instruction.Accounts)

				if len(accounts) == 0 {
					// NOTE: there must be at least 1 "valid" nullstring in the list since the record is `repeated` in BQ.
					accounts = append(accounts, "")
				}

				if len(paramRecords) == 0 {
					paramRecords = append(paramRecords, &pbcodegen.ParamsRecord{
						Key:   nil,
						Value: nil,
					})
				}

				index := int64(instruction_index)
				InstructionRecord := pbcodegen.InstructionRecord{
					BlockSlot:       &blockSlot,
					BlockTimestamp:  blockTimestamp,
					BlockHash:       &blockHash,
					TxSignature:     &txLeadSignature,
					Index:           &index,
					ParentIndex:     nil,
					Accounts:        accounts,
					Data:            instruction.Data,
					Program:         instruction.Program,
					ProgramId:       instruction.ProgramId,
					InstructionType: instruction_type,
					Params:          paramRecords,
					Parsed:          instruction_parsed,
				}

				InstructionRecords = append(InstructionRecords, &InstructionRecord)

				var tokenTransfer *TokenTransferData
				if instruction_type != nil {
					tokenTransfer = extractTokenTransfer(instruction.GetProgram(), *instruction_type, txLeadSignature, instruction_parsed_dict, prev_instruction)
				}

				if tokenTransfer != nil {
					var value *uint64
					if tokenTransfer.value != nil && *tokenTransfer.value != "" {
						_value, err := strconv.ParseUint(*tokenTransfer.value, 10, 64)
						if err != nil {
							fmt.Println("Error:", err)
							panic("terminating...")
						} else {
							value = &_value
						}
					}

					TokenTransferRecord := pbcodegen.TokenTransferRecord{
						BlockSlot:      &blockSlot,
						BlockTimestamp: blockTimestamp,
						BlockHash:      &blockHash,
						Source:         tokenTransfer.source,
						Destination:    tokenTransfer.destination,
						Authority:      tokenTransfer.authority,
						Value:          value,
						Decimals:       tokenTransfer.decimals,
						Fee:            tokenTransfer.fee,
						FeeDecimals:    tokenTransfer.feeDecimals,
						Memo:           tokenTransfer.memo,
						Mint:           tokenTransfer.mint,
						MintAuthority:  tokenTransfer.mintAuthority,
						TransferType:   tokenTransfer.transferType,
						TxSignature:    &txLeadSignature,
					}
					TokenTransferRecords = append(TokenTransferRecords, &TokenTransferRecord)
				}
				prev_instruction = instruction
			}
		}

		tx_meta := transaction.GetMeta()

		if tx_meta == nil {
			var preTokenBalances []*pbcodegen.TokenBalanceRecord
			preTokenBalance := pbcodegen.TokenBalanceRecord{
				AccountIndex: nil,
				Mint:         nil,
				Owner:        nil,
				Amount:       nil,
				Decimals:     nil,
			}
			preTokenBalances = append(preTokenBalances, &preTokenBalance)

			var postTokenBalances []*pbcodegen.TokenBalanceRecord
			postTokenBalance := pbcodegen.TokenBalanceRecord{
				AccountIndex: nil,
				Mint:         nil,
				Owner:        nil,
				Amount:       nil,
				Decimals:     nil,
			}
			postTokenBalances = append(postTokenBalances, &postTokenBalance)

			var balanceChanges []*pbcodegen.BalanceChangeRecord
			balanceChangeRecord := pbcodegen.BalanceChangeRecord{
				Account: nil,
				Before:  nil,
				After:   nil,
			}
			balanceChanges = append(balanceChanges, &balanceChangeRecord)

			lm := []string{""}

			txIndex := int64(transaction_index)
			TransactionRecord := pbcodegen.TransactionRecord{
				Signature:            &txLeadSignature,
				BlockHash:            &blockHash,
				RecentBlockHash:      recentBlockHash,
				BlockSlot:            &blockSlot,
				BlockTimestamp:       blockTimestamp,
				Index:                &txIndex,
				Fee:                  nil,
				Status:               nil,
				Err:                  nil,
				ComputeUnitsConsumed: nil,
				Accounts:             transactionAccountRecords,
				LogMessages:          lm,
				BalanceChanges:       balanceChanges,
				PreTokenBalances:     preTokenBalances,
				PostTokenBalances:    postTokenBalances,
			}
			TransactionRecords = append(TransactionRecords, &TransactionRecord)

		} else {
			fee := tx_meta.GetFee()
			tx_err := tx_meta.GetErr()
			var status string
			if tx_err == nil {
				status = "Success"
			} else {
				status = "Fail"
			}
			errString := tx_err.GetErr()

			lm := tx_meta.GetLogMessages()
			var log_messages []string
			if len(lm) > 0 {
				log_messages = append(log_messages, lm...)
			} else {
				log_messages = append(log_messages, "")
			}

			pre_balances := tx_meta.GetPreBalances()
			post_balances := tx_meta.GetPostBalances()

			var balanceChanges []*pbcodegen.BalanceChangeRecord
			for i, account := range transaction_accounts {
				if i >= len(pre_balances) || i >= len(post_balances) {
					break
				}
				pre_balance := pre_balances[i]
				post_balance := post_balances[i]
				pubkey := account.GetPubkey()
				balanceChange := pbcodegen.BalanceChangeRecord{
					Account: &pubkey,
					Before:  &pre_balance,
					After:   &post_balance,
				}
				balanceChanges = append(balanceChanges, &balanceChange)
			}

			var preTokenBalances []*pbcodegen.TokenBalanceRecord
			for _, preTokenBalance := range tx_meta.GetPreTokenBalances() {
				uiTokenAmount := preTokenBalance.GetUiTokenAmount()
				accountIndex := int64(preTokenBalance.AccountIndex)
				decimals := int64(uiTokenAmount.Decimals)
				preTokenBalanceRecord := pbcodegen.TokenBalanceRecord{
					AccountIndex: &accountIndex,
					Mint:         &preTokenBalance.Mint,
					Owner:        &preTokenBalance.Owner,
					Amount:       &uiTokenAmount.Amount,
					Decimals:     &decimals,
				}
				preTokenBalances = append(preTokenBalances, &preTokenBalanceRecord)
			}
			if len(preTokenBalances) == 0 {
				preTokenBalances = append(preTokenBalances, &pbcodegen.TokenBalanceRecord{AccountIndex: nil, Mint: nil, Owner: nil, Amount: nil, Decimals: nil})
			}

			var postTokenBalances []*pbcodegen.TokenBalanceRecord
			for _, postTokenBalance := range tx_meta.GetPostTokenBalances() {
				uiTokenAmount := postTokenBalance.GetUiTokenAmount()
				accountIndex := int64(postTokenBalance.AccountIndex)
				decimals := int64(uiTokenAmount.Decimals)
				postTokenBalanceRecord := pbcodegen.TokenBalanceRecord{
					AccountIndex: &accountIndex,
					Mint:         &postTokenBalance.Mint,
					Owner:        &postTokenBalance.Owner,
					Amount:       &uiTokenAmount.Amount,
					Decimals:     &decimals,
				}
				postTokenBalances = append(postTokenBalances, &postTokenBalanceRecord)
			}
			if len(postTokenBalances) == 0 {
				postTokenBalances = append(postTokenBalances, &pbcodegen.TokenBalanceRecord{AccountIndex: nil, Mint: nil, Owner: nil, Amount: nil, Decimals: nil})
			}

			if !tx_meta.GetInnerInstructionsNone() {
				for _, inner_instructions := range tx_meta.GetInnerInstructions() {
					var prev_inner_instruction *pbcodegen.InnerInstruction
					for j, inner_instruction := range inner_instructions.GetInstructions() {
						if inner_instruction == nil {
							continue
						}
						instruction_parsed_dict := inner_instruction.ParsedDict
						var instruction_parsed *string
						var instruction_type *string
						var param_records []*pbcodegen.ParamsRecord
						if instruction_parsed_dict == nil {
							instruction_type = nil
							instruction_parsed = inner_instruction.ParsedString
							param_record := pbcodegen.ParamsRecord{
								Key:   nil,
								Value: nil,
							}
							param_records = append(param_records, &param_record)

						} else {
							instruction_type = instruction_parsed_dict.Type
							instruction_parsed = instruction_parsed_dict.Parsed

							param_keys_values := instruction_parsed_dict.Info
							if param_keys_values == nil {
								param_records = append(param_records, &pbcodegen.ParamsRecord{
									Key:   nil,
									Value: nil,
								})
							} else {
								var keys_values_map map[string]interface{}
								err := json.Unmarshal([]byte(*param_keys_values), &keys_values_map)
								if err != nil {
									fmt.Println("Error:", err)
									panic("Unexpected instruction params value")
								}
								for k, v := range keys_values_map {
kCopy := k // Create a copy of k
									jsonBytes, err := json.Marshal(v)
									if err != nil {
										panic(err)
									}
									jsonString := string(jsonBytes)

//fmt.Println("the inner k is:", k)
//fmt.Println("the inner v string is:", jsonString)
									param_record := pbcodegen.ParamsRecord{
										Key:   &kCopy,
										Value: &jsonString,
									}
//fmt.Println("the inner param_record is:", param_record)					
				param_records = append(param_records, &param_record)
								}
							}
						}

						var accounts []string
						if len(inner_instruction.Accounts) > 0 {
							accounts = append(accounts, inner_instruction.Accounts...)
						} else {
							accounts = append(accounts, "")
						}

						innerIndex := int64(j)
						parentIndex := int64(inner_instructions.Index)
						InstructionRecord := pbcodegen.InstructionRecord{
							BlockSlot:       &blockSlot,
							BlockTimestamp:  blockTimestamp,
							BlockHash:       &blockHash,
							TxSignature:     &txLeadSignature,
							Index:           &innerIndex,
							ParentIndex:     &parentIndex,
							Accounts:        accounts,
							Data:            inner_instruction.Data,
							Program:         inner_instruction.Program,
							ProgramId:       inner_instruction.ProgramId,
							InstructionType: instruction_type,
							Params:          param_records,
							Parsed:          instruction_parsed,
						}

						InstructionRecords = append(InstructionRecords, &InstructionRecord)
						var tokenTransfer *TokenTransferData
						if instruction_type != nil {
							tokenTransfer = extractTokenTransfer(inner_instruction.GetProgram(), *instruction_type, txLeadSignature, instruction_parsed_dict, prev_inner_instruction)
						}

						if tokenTransfer != nil {
							var value *uint64
							if tokenTransfer.value != nil && *tokenTransfer.value != "" {
								_value, err := strconv.ParseUint(*tokenTransfer.value, 10, 64)
								if err != nil {
									fmt.Println("Error:", err)
									panic("terminating...")
								} else {
									value = &_value
								}
							}
							tokenTransfer_record := pbcodegen.TokenTransferRecord{
								BlockSlot:      &blockSlot,
								BlockTimestamp: blockTimestamp,
								BlockHash:      &blockHash,
								Source:         tokenTransfer.source,
								Destination:    tokenTransfer.destination,
								Authority:      tokenTransfer.authority,
								Value:          value,
								Decimals:       tokenTransfer.decimals,
								Mint:           tokenTransfer.mint,
								MintAuthority:  tokenTransfer.mintAuthority,
								TransferType:   tokenTransfer.transferType,
								TxSignature:    &txLeadSignature,
							}
							TokenTransferRecords = append(TokenTransferRecords, &tokenTransfer_record)
						}
						prev_inner_instruction = inner_instruction
					}

				}
			}

			txIndex := int64(transaction_index)
			TransactionRecord := pbcodegen.TransactionRecord{
				BlockSlot:            &blockSlot,
				BlockTimestamp:       blockTimestamp,
				BlockHash:            &blockHash,
				Signature:            &txLeadSignature,
				RecentBlockHash:      recentBlockHash,
				Index:                &txIndex,
				Fee:                  &fee,
				Status:               &status,
				Err:                  &errString,
				ComputeUnitsConsumed: tx_meta.ComputeUnitsConsumed,
				Accounts:             transactionAccountRecords,
				LogMessages:          log_messages,
				BalanceChanges:       balanceChanges,
				PreTokenBalances:     preTokenBalances,
				PostTokenBalances:    postTokenBalances,
			}
			TransactionRecords = append(TransactionRecords, &TransactionRecord)
		}
	}
	return TransactionRecords, InstructionRecords, TokenTransferRecords
}
